spring-integration-core/src/main/java/org/springframework/integration/support/json/MutableMessageJacksonDeserializer.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.support.json;
+
+import java.io.IOException;
+
+import org.springframework.integration.support.MutableMessage;
+import org.springframework.integration.support.MutableMessageHeaders;
+import org.springframework.util.ClassUtils;
+
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonNode;
+
+/**
+ * The {@link MessageJacksonDeserializer} implementation for the {@link MutableMessage}.
+ *
+ * @author Artem Bilan
+ *
+ * @since 5.0
+ */
+public class MutableMessageJacksonDeserializer extends MessageJacksonDeserializer<MutableMessage<?>> {
```

Comment:
So, then may be it's time to consider deprecating `MutableMessage` and gradually remove all references to it internally. . . (as a separate issue)